libp2p/pubsub/gossipsub.py

```diff
@@ -204,42 +215,46 @@ def _get_peers_to_send(
         :return: a generator of the peer ids who we send data to.
         """
-        send_to: Set[ID] = set()
+        send_to: List[ID] = []
         for topic in topic_ids:
             if topic not in self.pubsub.peer_topics:
                 continue
 
             # floodsub peers
-            for peer_id in self.pubsub.peer_topics[topic]:
-                # FIXME: `gossipsub.peers_floodsub` can be changed to `gossipsub.peers` in go.
-                #   This will improve the efficiency when searching for a peer's protocol id.
-                if peer_id in self.peers_floodsub:
-                    send_to.add(peer_id)
+            # FIXME: `gossipsub.peers_floodsub` can be changed to `gossipsub.peers` in go.
+            #   This will improve the efficiency when searching for a peer's protocol id.
+            floodsub_peers: List[ID] = [
+                peer_id
+                for peer_id in self.pubsub.peer_topics[topic]
+                if peer_id in self.peers_floodsub
+            ]
 
             # gossipsub peers
-            in_topic_gossipsub_peers: List[ID] = None
-            # TODO: Do we need to check `topic in self.pubsub.my_topics`?
+            gossipsub_peers: List[ID] = []
             if topic in self.mesh:
-                in_topic_gossipsub_peers = self.mesh[topic]
+                gossipsub_peers = self.mesh[topic]
             else:
-                # TODO(robzajac): Is topic DEFINITELY supposed to be in fanout if we are not
-                #   subscribed?
-                # I assume there could be short periods between heartbeats where topic may not
-                # be but we should check that this path gets hit appropriately
-
-                if (topic not in self.fanout) or (len(self.fanout[topic]) == 0):
-                    # If no peers in fanout, choose some peers from gossipsub peers in topic.
-                    self.fanout[topic] = self._get_in_topic_gossipsub_peers_from_minus(
-                        topic, self.degree, []
-                    )
-                in_topic_gossipsub_peers = self.fanout[topic]
-            for peer_id in in_topic_gossipsub_peers:
-                send_to.add(peer_id)
+                # When we publish to a topic that we have not subscribe to, we randomly pick
+                # `self.degree` number of peers who have subscribed to the topic and add them
+                # as our `fanout` peers.
+                topic_in_fanout: bool = topic in self.fanout
+                fanout_peers: List[ID] = self.fanout[topic] if topic_in_fanout else []
+                fanout_size = len(fanout_peers)
+                if not topic_in_fanout or (
+                    topic_in_fanout and fanout_size < self.degree
+                ):
+                    if topic in self.pubsub.peer_topics:
+                        # Combine fanout peers with selected peers
+                        fanout_peers += self._get_in_topic_gossipsub_peers_from_minus(
```

Comment:
Note: one thing I'm thinking is to have methods like `_get_fanout_peers(topic)`, `_get_mesh_peers(topic)`, and `_get_ peers_to_emit_ihave_to`, to make the code more readable.